![text here](./images/moneyball_title.jpg)
______

# Synopsis

## Background
_Source: Wikipedia_

It was the 102nd season in franchise history. The Athletics finished first in the American League West with a record of 103-59.

The Athletics' 2002 campaign ranks among the most famous in franchise history. Following the 2001 season, Oakland saw the departure of three key players. Billy Beane, the team's general manager, responded with a series of under-the-radar free agent signings. The new-look Athletics, despite a comparative lack of star power, surprised the baseball world by besting the 2001 team's regular season record. The team is most famous, however, for winning 20 consecutive games between August 13 and September 4, 2002. The Athletics' season was the subject of Michael Lewis's 2003 book Moneyball: The Art of Winning an Unfair Game (Lewis was given the opportunity to follow the team around throughout the season). A film adaptation of the book, also titled Moneyball, was released in 2011 starring _Brad Pitt and Jonah Hill_.

## Moneyball book
[_Moneyball: The art of winning an unfair game by Micheal Lewis_](https://www.amazon.com/Moneyball-Art-Winning-Unfair-Game-ebook/dp/B000RH0C8G/ref=sr_1_2?ie=UTF8&qid=1546997969&sr=8-2&keywords=moneyball)

The central premise of Moneyball is that the collective wisdom of baseball insiders (including players, managers, coaches, scouts, and the front office) over the past century is subjective and often flawed. Statistics such as stolen bases, runs batted in, and batting average, typically used to gauge players, are relics of a 19th-century view of the game and the statistics available at that time. Before sabermetrics was introduced to baseball, teams were dependent on the skills of their scouts to find and evaluate players. Scouts are experienced in the sport, usually having been players or coaches.[1] The book argues that the Oakland A's' front office took advantage of more analytical gauges of player performance to field a team that could better compete against richer competitors in Major League Baseball (MLB).

Rigorous statistical analysis had demonstrated that on-base percentage and slugging percentage are better indicators of offensive success, and the A's became convinced that these qualities were cheaper to obtain on the open market than more historically valued qualities such as speed and contact. These observations often flew in the face of conventional baseball wisdom and the beliefs of many baseball scouts and executives.

By re-evaluating their strategy in this way, the 2002 Athletics, with approximately \44 million in salary, were competitive with larger market teams such as the New York Yankees, who spent over \$125 million in payroll that season. Because of its smaller budget, Oakland had to find players undervalued by the market, and their system has proven itself thus far. The approach brought the A's to the playoffs in 2002 and 2003.

<img src="images/salary.png" width="600">

Because of the team's smaller revenues, Oakland is forced to find players undervalued by the market, and their system for finding value in undervalued players has proven itself thus far. This approach brought the A's to the playoffs in 2002 and 2003.

In this project we work with some data and with the goal of trying to find replacement players for the ones lost at the start of the off-season - During the 2001–02 offseason, the team lost three key free agents to larger market teams: 2000 AL MVP Jason Giambi to the New York Yankees, outfielder Johnny Damon to the Boston Red Sox, and closer Jason Isringhausen to the St. Louis Cardinals.

## Data Source 
From [Sean Lahaman's](http://www.seanlahman.com/baseball-archive/statistics/) website. The metadata for the csv files are located **readme2013.txt** file.

### Let us first load the data and look at the head of the file

In [1]:
batting <- read.csv('./data/Batting.csv')

In [2]:
head(batting, 3)

playerID  yearID stint teamID lgID G  G_batting AB R H ⋯ SB CS BB SO IBB HBP
1 aardsda01 2004   1     SFN    NL   11 11        0  0 0 ⋯ 0  0  0  0  0   0  
2 aardsda01 2006   1     CHN    NL   45 43        2  0 0 ⋯ 0  0  0  0  0   0  
3 aardsda01 2007   1     CHA    AL   25  2        0  0 0 ⋯ 0  0  0  0  0   0  
  SH SF GIDP G_old
1 0  0  0    11   
2 1  0  0    45   
3 0  0  0     2

### Structure of our dataset?

In [3]:
str(batting)

'data.frame':	97889 obs. of  24 variables:
 $ playerID : Factor w/ 18107 levels "aardsda01","aaronha01",..: 1 1 1 1 1 1 1 2 2 2 ...
 $ yearID   : int  2004 2006 2007 2008 2009 2010 2012 1954 1955 1956 ...
 $ stint    : int  1 1 1 1 1 1 1 1 1 1 ...
 $ teamID   : Factor w/ 149 levels "ALT","ANA","ARI",..: 117 35 33 16 116 116 93 80 80 80 ...
 $ lgID     : Factor w/ 6 levels "AA","AL","FL",..: 4 4 2 2 2 2 2 4 4 4 ...
 $ G        : int  11 45 25 47 73 53 1 122 153 153 ...
 $ G_batting: int  11 43 2 5 3 4 NA 122 153 153 ...
 $ AB       : int  0 2 0 1 0 0 NA 468 602 609 ...
 $ R        : int  0 0 0 0 0 0 NA 58 105 106 ...
 $ H        : int  0 0 0 0 0 0 NA 131 189 200 ...
 $ X2B      : int  0 0 0 0 0 0 NA 27 37 34 ...
 $ X3B      : int  0 0 0 0 0 0 NA 6 9 14 ...
 $ HR       : int  0 0 0 0 0 0 NA 13 27 26 ...
 $ RBI      : int  0 0 0 0 0 0 NA 69 106 92 ...
 $ SB       : int  0 0 0 0 0 0 NA 2 3 2 ...
 $ CS       : int  0 0 0 0 0 0 NA 2 1 4 ...
 $ BB       : int  0 0 0 0 0 0 NA 28 49 37 ...
 $ S

In [4]:
head(batting$AB)

[1] 0 2 0 1 0 0

In [5]:
head(batting$X2B)

[1] 0 0 0 0 0 0

## Feature Engineering

We need to add a couple more statistics that were used in Moneyball. Thesea are:

- [Batting Average](https://en.wikipedia.org/wiki/Batting_average)
- [On Base Percentage](https://en.wikipedia.org/wiki/On-base_percentage)
- [Slugging Percentage](https://en.wikipedia.org/wiki/Slugging_percentage)

In [6]:
# Batting Average
batting$BA <- batting$H / batting$AB

In [7]:
tail(batting$BA, 5)

[1] 0.1230769 0.2746479 0.1470588 0.2745098 0.2138728

In [8]:
# On base percentage
batting$OBP <- (batting$H + batting$BB + batting$HBP) / (batting$AB + batting$BB + batting$HBP + batting$SF)

### Notice that singles (1B) are are not in the dataframe we need to calculate it. But that's dasy Since 
$$ 1B = H - 2B - 3B - HR $$

In [9]:
batting$X1B = batting$H - batting$X2B - batting$X3B - batting$HR

In [10]:
#  Slugging average
batting$SLG <- (batting$X1B + 2 * batting$X2B + 3 * batting$X3B + 4 * batting$HR) / batting$AB

### How does the DF look now?

In [11]:
str(batting)

'data.frame':	97889 obs. of  28 variables:
 $ playerID : Factor w/ 18107 levels "aardsda01","aaronha01",..: 1 1 1 1 1 1 1 2 2 2 ...
 $ yearID   : int  2004 2006 2007 2008 2009 2010 2012 1954 1955 1956 ...
 $ stint    : int  1 1 1 1 1 1 1 1 1 1 ...
 $ teamID   : Factor w/ 149 levels "ALT","ANA","ARI",..: 117 35 33 16 116 116 93 80 80 80 ...
 $ lgID     : Factor w/ 6 levels "AA","AL","FL",..: 4 4 2 2 2 2 2 4 4 4 ...
 $ G        : int  11 45 25 47 73 53 1 122 153 153 ...
 $ G_batting: int  11 43 2 5 3 4 NA 122 153 153 ...
 $ AB       : int  0 2 0 1 0 0 NA 468 602 609 ...
 $ R        : int  0 0 0 0 0 0 NA 58 105 106 ...
 $ H        : int  0 0 0 0 0 0 NA 131 189 200 ...
 $ X2B      : int  0 0 0 0 0 0 NA 27 37 34 ...
 $ X3B      : int  0 0 0 0 0 0 NA 6 9 14 ...
 $ HR       : int  0 0 0 0 0 0 NA 13 27 26 ...
 $ RBI      : int  0 0 0 0 0 0 NA 69 106 92 ...
 $ SB       : int  0 0 0 0 0 0 NA 2 3 2 ...
 $ CS       : int  0 0 0 0 0 0 NA 2 1 4 ...
 $ BB       : int  0 0 0 0 0 0 NA 28 49 37 ...
 $ S

In [12]:
sal <- read.csv('./data/Salaries.csv')

In [13]:
summary(sal)

     yearID         teamID      lgID            playerID         salary        
 Min.   :1985   CLE    :  867   AL:11744   moyerja01:   25   Min.   :       0  
 1st Qu.:1993   LAN    :  861   NL:12212   vizquom01:   24   1st Qu.:  250000  
 Median :1999   PHI    :  861              glavito02:   23   Median :  507950  
 Mean   :1999   SLN    :  858              bondsba01:   22   Mean   : 1864357  
 3rd Qu.:2006   BAL    :  855              griffke02:   22   3rd Qu.: 2100000  
 Max.   :2013   NYA    :  855              thomeji01:   22   Max.   :33000000  
                (Other):18799              (Other)  :23818                     

In [14]:
head(sal, 4)

yearID teamID lgID playerID  salary 
1 1985   BAL    AL   murraed02 1472819
2 1985   BAL    AL   lynnfr01  1090000
3 1985   BAL    AL   ripkeca01  800000
4 1985   BAL    AL   lacyle01   725000

In [15]:
summary(batting)

      playerID         yearID         stint           teamID        lgID      
 mcguide01:   31   Min.   :1871   Min.   :1.000   CHN    : 4720   AA  : 1890  
 henderi01:   29   1st Qu.:1931   1st Qu.:1.000   PHI    : 4621   AL  :44369  
 newsobo01:   29   Median :1970   Median :1.000   PIT    : 4575   FL  :  470  
 johnto01 :   28   Mean   :1962   Mean   :1.077   SLN    : 4535   NL  :49944  
 kaatji01 :   28   3rd Qu.:1995   3rd Qu.:1.000   CIN    : 4393   PL  :  147  
 ansonca01:   27   Max.   :2013   Max.   :5.000   CLE    : 4318   UA  :  332  
 (Other)  :97717                                  (Other):70727   NA's:  737  
       G            G_batting            AB              R         
 Min.   :  1.00   Min.   :  0.00   Min.   :  0.0   Min.   :  0.00  
 1st Qu.: 13.00   1st Qu.:  7.00   1st Qu.:  9.0   1st Qu.:  0.00  
 Median : 35.00   Median : 32.00   Median : 61.0   Median :  5.00  
 Mean   : 51.65   Mean   : 49.13   Mean   :154.1   Mean   : 20.47  
 3rd Qu.: 81.00   3rd Qu.: 8

Note that **batting** data contain data all the way back to 1871. We only need need data after 1985 so that we may merge these two dataframes in order for our analysis to proceed. We re-assign batting using the _subset_ function in R

In [16]:
batting <- subset(batting, yearID >= 1985)

In [17]:
summary(batting)

      playerID         yearID         stint          teamID      lgID      
 moyerja01:   27   Min.   :1985   Min.   :1.00   SDN    : 1313   AA:    0  
 mulhote01:   26   1st Qu.:1993   1st Qu.:1.00   CLE    : 1306   AL:17226  
 weathda01:   26   Median :2000   Median :1.00   PIT    : 1299   FL:    0  
 maddugr01:   25   Mean   :2000   Mean   :1.08   NYN    : 1297   NL:18426  
 sierrru01:   25   3rd Qu.:2007   3rd Qu.:1.00   BOS    : 1279   PL:    0  
 thomeji01:   25   Max.   :2013   Max.   :4.00   CIN    : 1279   UA:    0  
 (Other)  :35498                                 (Other):27879             
       G           G_batting            AB              R         
 Min.   :  1.0   Min.   :  0.00   Min.   :  0.0   Min.   :  0.00  
 1st Qu.: 14.0   1st Qu.:  4.00   1st Qu.:  3.0   1st Qu.:  0.00  
 Median : 34.0   Median : 27.00   Median : 47.0   Median :  4.00  
 Mean   : 51.7   Mean   : 46.28   Mean   :144.7   Mean   : 19.44  
 3rd Qu.: 77.0   3rd Qu.: 77.00   3rd Qu.:241.0   3rd Qu.

In [18]:
combo <- merge(batting, sal, by=c('playerID', 'yearID'))

In [19]:
summary(combo)

      playerID         yearID         stint          teamID.x     lgID.x    
 moyerja01:   27   Min.   :1985   Min.   :1.000   LAN    :  940   AA:    0  
 thomeji01:   25   1st Qu.:1993   1st Qu.:1.000   PHI    :  937   AL:12292  
 weathda01:   25   Median :1999   Median :1.000   BOS    :  935   FL:    0  
 vizquom01:   24   Mean   :1999   Mean   :1.098   NYA    :  928   NL:13105  
 gaettga01:   23   3rd Qu.:2006   3rd Qu.:1.000   CLE    :  920   PL:    0  
 griffke02:   23   Max.   :2013   Max.   :4.000   SDN    :  914   UA:    0  
 (Other)  :25250                                  (Other):19823             
       G            G_batting            AB              R         
 Min.   :  1.00   Min.   :  0.00   Min.   :  0.0   Min.   :  0.00  
 1st Qu.: 26.00   1st Qu.:  8.00   1st Qu.:  5.0   1st Qu.:  0.00  
 Median : 50.00   Median : 42.00   Median : 85.0   Median :  9.00  
 Mean   : 64.06   Mean   : 57.58   Mean   :182.4   Mean   : 24.71  
 3rd Qu.:101.00   3rd Qu.:101.00   3rd Qu.:3

### Find the lost players by subsetting. 
As previously mentioned , the Oakland A's lost 3 key players during off-season. We want to get their statistics to see what we have to replace. The players lost were:

- Frist Baseman 2000 AL MVP Jason Gambi (giambja01) to the New York Yankees
- Outfielder Johnny Damon (damonjo01) to the Boston Red Sox
- Infielder Rainer Gustavo "Ray" Olmedo (saenol01)

In [20]:
lost_players <- subset(combo, playerID %in% c('giambja01', 'damonjo01', 'saenzol01'))

In [21]:
head(lost_players)

playerID  yearID stint teamID.x lgID.x G   G_batting AB  R   H   ⋯ SF GIDP
5135 damonjo01 1995   1     KCA      AL      47  47       188  32  53 ⋯  3  2  
5136 damonjo01 1996   1     KCA      AL     145 145       517  61 140 ⋯  5  4  
5137 damonjo01 1997   1     KCA      AL     146 146       472  70 130 ⋯  1  3  
5138 damonjo01 1998   1     KCA      AL     161 161       642 104 178 ⋯  3  4  
5139 damonjo01 1999   1     KCA      AL     145 145       583 101 179 ⋯  4 13  
5140 damonjo01 2000   1     KCA      AL     159 159       655 136 214 ⋯ 12  7  
     G_old BA        OBP       X1B SLG       teamID.y lgID.y salary 
5135  47   0.2819149 0.3235294  34 0.4414894 KCA      AL      109000
5136 145   0.2707930 0.3129496 107 0.3675048 KCA      AL      180000
5137 146   0.2754237 0.3378378 102 0.3855932 KCA      AL      240000
5138 161   0.2772586 0.3394625 120 0.4392523 KCA      AL      460000
5139 145   0.3070326 0.3789954 117 0.4768439 KCA      AL     2100000
5140 159   0.3267176 0.3819918 146 0.4946565 KCA      AL     4000000

Since all of these players were lost after 2001 in the off season, let us only concern ourselves with data from 2001.

In [22]:
lost_players <- subset(lost_players, yearID %in% 2001)

In [23]:
lost_players 

playerID  yearID stint teamID.x lgID.x G   G_batting AB  R   H   ⋯ SF
5141  damonjo01 2001   1     OAK      AL     155 155       644 108 165 ⋯ 4 
7878  giambja01 2001   1     OAK      AL     154 154       520 109 178 ⋯ 9 
20114 saenzol01 2001   1     OAK      AL     106 106       305  33  67 ⋯ 3 
      GIDP G_old BA        OBP       X1B SLG       teamID.y lgID.y salary 
5141   7   155   0.2562112 0.3235294 118 0.3633540 OAK      AL     7100000
7878  17   154   0.3423077 0.4769001  91 0.6596154 OAK      AL     4103333
20114  9   106   0.2196721 0.2911765  36 0.3836066 OAK      AL      290000

Let us go ahead and drop the unncessary columns from this dataset. Specifically we only want the following columns:

- playerID
- H
- X2B
- X3B
- HR
- OBP
- SLG
- BA
- AB

In [24]:
# Create a list of columns to keep
keeps <- c('playerID', 'H', 'X2B', 'X3B', 'HR', 'OBP', 'SLG', 'BA', 'AB')

In [25]:
lost_players <- subset(lost_players, select = keeps)

In [26]:
lost_players

playerID  H   X2B X3B HR OBP       SLG       BA        AB 
5141  damonjo01 165 34  4    9 0.3235294 0.3633540 0.2562112 644
7878  giambja01 178 47  2   38 0.4769001 0.6596154 0.3423077 520
20114 saenzol01  67 21  1    9 0.2911765 0.3836066 0.2196721 305

## Replacement Players

Now our task is to find replacment players for those that we have lost. Subjected to these constrains of course:

- The total combined salary of the three players cannot exceed 15 million dollars.
- Their main combined number of at bats (AB) needs to be equal to or greater than the lost players
- Their mean OBP had to equal to or greater than the mean OBP of lost players.

Let us first compute the combined at bats (AB) of lost players and their mean OBP. These are to be used
as threasholds in our player replacement scheme. 

In [27]:
# Lost players' combined AB
lp_combined_AB = sum(lost_players$AB)
lp_combined_AB

[1] 1469

In [28]:
# lost players' mean OBP
lp_mean_OBP = mean(lost_players$OBP)
lp_mean_OBP

[1] 0.3638687

In [29]:
salary_cap = 15000000

In [30]:
combo_2001 = subset(combo, yearID %in% 2001)

In [31]:
combo_keeps = c('playerID', 'H', 'X2B', 'X3B', 'HR', 'OBP', 'SLG', 'BA', 'AB', 'salary')

In [32]:
combo_2001 = subset(combo_2001, select = combo_keeps)

In [111]:
head(combo_2001)

playerID  H   X2B X3B HR OBP       SLG        BA         AB  salary 
17  abbotje01  11  3  0    0 0.3260870 0.33333333 0.26190476  42  300000
37  abbotku01   2  0  0    0 0.2222222 0.22222222 0.22222222   9  600000
44  abbotpa01   1  0  0    0 0.2500000 0.25000000 0.25000000   4 1700000
62  abreubo01 170 48  4   31 0.3934659 0.54251701 0.28911565 588 4983000
130 adamste01   2  1  0    0 0.1162791 0.07692308 0.05128205  39 2600000
157 agbaybe01  82 14  2    6 0.3639053 0.39864865 0.27702703 296  260000


# Possible solution outlines

One way to proceed is to use 'groupby' statements to group these players by the various constraints and then with a bit of dumb luck we will be able to narrow down the choices for a possible combination of three players. But in here let's try to find _all_ of the possible combinations of players that satisfy the given constraints. This type of a problem is what is known as _constraint programming_ or a _constraint satisfaction problem_.
 
## Options 1 (Brute Force):

The obvious way is to find all possible combinations of players taken 3 at a time and the filter out the unwanted entries according to the given constraints. But note that this produces $$C(915, 3) = 127,258,505$$ possible combinations. This is obviously not a very elegent solution as we will quicly run out of memory to hold these combintations. This is certainly not scalable.


## Options 2: (Using integer programming)

We can use the [*LpSolve*](https://cran.r-project.org/web/packages/lpSolve) library to solve the problem. Documentation as well as an example of a solution of a sample problem may be found [here](https://cran.r-project.org/web/packages/lpSolve/lpSolve.pdf)

### Set up of the problem.


We want to find exactly three players subjected to the above constraints. Let $x, y$ and $z$ be three boolean masks of the players. For example if `abbotje01` the first player is in the team then his boolean mask will be of the form 

$$
x = [1, 0, 0, \cdots, 0,0]
$$

where $x$ is a vector of length 915 (the total number of player available for selection). Our integer programming
problem is now of the following form.

Find optimal $x,y$ and $z$ subjected to

$$
x + y + z = 3, \\
x\$salary + y\$salary + z\$salary \leq 15e6, \\
x\$AB + y\$AB + z\$AB \geq 1469, \\
x\$OBP + y\$OBP + z\$OBP \geq 3 * 0.3638
$$

## Solution

In [36]:
library(lpSolve)

In [37]:
# Create the objective function
f.obj <- rep(1, nrow(combo_2001))

In [112]:
f.con <-
    matrix(c(f.obj,
             as.vector(combo_2001$salary),
             as.vector(combo_2001$AB),
             as.vector(combo_2001$OBP)), nrow=4, byrow=TRUE)

# whether equal, greter than or less than
f.dir <- c('==', '<=', '>=', '>=')


# Create the right hand sice of the contraints.
f.rhs <- c(3, # number of players
           salary_cap, # max combined salary
           lp_combined_AB, # minimum combined AB
           3 * lp_mean_OBP # minimum combined mean OPB
          )

In [113]:
mylp <- lp ("max", f.obj, f.con, f.dir, f.rhs, all.bin=TRUE, num.bin.solns = 20)
numcols <- 915
numsols <- mylp$num.bin.solns

solutions <- matrix(head(mylp$solution, numcols*numsols), nrow=numsols, byrow=TRUE)

In the above `solutions` is a $915 \times 20$ matrix in which each row is a possible solution to the problem. We have decided to find only $20$ solutions, but that number may be increased as desired. For examaple if a solution
is of the form

$$
[0, 1, 1, 0 , 0, 1, 0, 0, \cdots, 0]
$$

It means that the player combo of the 2nd, 3rd and 6th players in the `combo_2001` data frame may be used as replacments for the lost players. Here is the possible player combo, by choosing the 10th solution of the solutions set. There is no particular reason why we chose 10. Any number chosen from 1 through 20 will equally well work.

In [114]:
player_combo <- combo_2001[which(!solutions[10,]==0),]
player_combo

playerID  H   X2B X3B HR OBP       SLG       BA        AB  salary  
6263  eckstda01 166 26   2   4 0.3549383 0.3573883 0.2852234 582   200000
19766 rolliji01 180 29  12  14 0.3234880 0.4192073 0.2743902 656   200000
23777 walkela01 174 35   3  38 0.4492512 0.6619718 0.3501006 497 12166667

## But does this solution hold water?

Let us check if the conditions are indeed satisfied by this possible solution.

In [85]:
mean(player_combo$OBP)

[1] 0.3758925

In [86]:
sum(player_combo$AB)

[1] 1735

In [87]:
sum(player_combo$salary)

[1] 12566667

They are indeed satisfied. Here is another possibility.

In [115]:
player_combo1 <- combo_2001[which(!solutions[3,]==0),]
player_combo1

playerID  H   X2B X3B HR OBP       SLG       BA        AB  salary  
2277  bondsba01 156 32   2  73 0.5150602 0.8634454 0.3277311 476 10300000
6263  eckstda01 166 26   2   4 0.3549383 0.3573883 0.2852234 582   200000
19766 rolliji01 180 29  12  14 0.3234880 0.4192073 0.2743902 656   200000

In [116]:
mean(player_combo1$OBP)

[1] 0.3978289

In [117]:
sum(player_combo1$AB)

[1] 1714

In [118]:
sum(player_combo1$salary)

[1] 10700000

## How about an optimal solution from amongst all of the possible solutions?.

That can indeed be achieved easily by setting the `all.int=TRUE` parameter and getting rid of the `all.bin=TRUE` parameters. The `num.bin.solns`  parameter is redundant in this case.

In [120]:
optimal_lp <- lp ("max", 
                  f.obj, 
                  f.con, 
                  f.dir, 
                  f.rhs, 
                  all.int=TRUE # can't have half a player or 0.9 of a player
                 )

In [123]:
# sanity check
length(optimal_lp$solution)

[1] 915

In [124]:
opt_sol = optimal_lp$solution

In [125]:
optimal_combo <- combo_2001[which(!opt_sol==0),]
optimal_combo

playerID  H   X2B X3B HR OBP       SLG       BA        AB  salary  
2277  bondsba01 156 32   2  73 0.5150602 0.8634454 0.3277311 476 10300000
18471 pujolal01 194 47   4  37 0.4029630 0.6101695 0.3288136 590   200000
19766 rolliji01 180 29  12  14 0.3234880 0.4192073 0.2743902 656   200000

## Check if the solution satisfies the constraints.

In [108]:
sum(optimal_combo$salary)

[1] 10700000

In [109]:
mean(optimal_combo$OBP)

[1] 0.4138371

In [110]:
sum(optimal_combo$AB)

[1] 1722

This optimal solution is probably impractical since Barry Bonds would never have been sold by the Giants. His ego was probably too big for him to have played for the Oakland A's anyway. :)